### PR TITLE
feat: extra memory regulation layer, revisit termination routines

### PR DESCRIPTION
**Pull request by nyannyacha**
## What kind of change does this PR introduce?

Feature

## Description

This PR introduces an additional layer of tracking and regulating the memory usage used by individual isolates using wakers and timers and garbage collection callback.

V8 uses page allocators to allocate linear memory for web assemblies, so there is no public API in `V8::Isolate` to detect allocations and deallocations.

Furthermore, it turns out that Deno sometimes wraps rust's buffers in `BackingStore` and attaches directly to an isolate, and there is no way for edge-runtime to trap this without a patch.

For the downstream edge-runtime to detect this and regulate excessive memory usage of each isolate, it will need to implement a memory usage tracking routine that is as cost-effective as possible.

This PR is the first plan for that.

### Other things

1. During iterative integration testing, I found an issue with the graceful shutdown routine and it was revisited. 
graceful shutting down routine is related to connection draining (#284) and additional PR will be submitted in the future.

2. Allowed some CLI arguments to be accepted as environment variables. (see [1f32918](https://github.com/supabase/edge-runtime/pull/286/commits/1f32918e0d9d888a4ad049112f53257987c91f74))

### CODE REVIEW

File: crates/base/src/deno_runtime.rs [hunk 1]
```diff
@@ -389,12 +468,25 @@ impl DenoRuntime {
             );
         }
 
+        if is_user_worker {
+            js_runtime.v8_isolate().add_gc_prologue_callback(
```
Comment (laktek) [inline crates/base/src/deno_runtime.rs:472]: @nyannyacha This callback would run before garbage collector would start?

File: crates/base/src/deno_runtime.rs [hunk 2]
```diff
@@ -435,15 +527,44 @@ impl DenoRuntime {
             .load_main_module(&main_module_url, mod_code)
             .await?;
 
+        if is_user_worker {
+            drop(rt::SUPERVISOR_RT.spawn({
+                let drop_token = mem_check_state.drop_token.clone();
+                let waker = mem_check_state.waker.clone();
+
+                async move {
+                    // TODO(Nyannyacha): Should we introduce exponential
```
Comment (laktek) [inline crates/base/src/deno_runtime.rs:536]: yeah, that'd be neat! Can be a follow up PR?

Comment (nyannyacha) [reply crates/base/src/deno_runtime.rs:536]: @laktek 
Planning to submit a PR in the next iteration.
In my own degradation experiments (see slack), I didn't find any significant degradation, so I just left it as a comment in this PR.

File: crates/base/src/deno_runtime.rs (hunk 1 above)
Comment (nyannyacha) [reply crates/base/src/deno_runtime.rs:472]: @laktek 
This callback is called just before V8 attempts to garbage collect.[^1][^2]

[^1]:https://source.chromium.org/chromium/chromium/src/+/main:v8/src/heap/heap.cc;l=1774?q=heap.cc
[^2]:https://source.chromium.org/chromium/chromium/src/+/main:v8/src/heap/heap.cc;l=1833?q=heap.cc

**Review by laktek — APPROVED** [review]
LGTM! Let's wait for @andreespirela to review it as well.

**Review by andreespirela — APPROVED** [review]
Lgtm

**Comment by nyannyacha** [comment]
wait

**Pull request merged**